modules/porous_flow/src/kernels/PorousFlowDarcyBase.C

```diff
@@ -93,22 +100,53 @@ PorousFlowDarcyBase::computeQpResidual()
 void
 PorousFlowDarcyBase::computeResidual()
 {
+  computeProtoFlux();
+  if (_pf_count_used)
+  {
+    unsigned pf_count(_pf_count->maxUpDownSwaps(0));
+    Moose::out << "elem = " << _current_elem->id() << "pf count returns " << pf_count << "\n";
+  }
   upwind(JacRes::CALCULATE_RESIDUAL, 0.0);
 }
 
 void
 PorousFlowDarcyBase::computeJacobian()
 {
+  // Compute the fluxes the mobility terms because we need to know which
+  // nodes are upwind and which are downwind
+  if (!_pf_count_used)
+    computeProtoFlux();
   upwind(JacRes::CALCULATE_JACOBIAN, _var.number());
 }
 
 void
 PorousFlowDarcyBase::computeOffDiagJacobian(unsigned int jvar)
 {
+  // Compute the fluxes the mobility terms because we need to know which
+  // nodes are upwind and which are downwind
+  if (!_pf_count_used)
+    computeProtoFlux();
   upwind(JacRes::CALCULATE_JACOBIAN, jvar);
 }
 
 void
+PorousFlowDarcyBase::computeProtoFlux()
+{
+  const unsigned int num_nodes = _test.size();
+  for (unsigned ph = 0; ph < _num_phases; ++ph)
+  {
+    _proto_flux[ph].assign(num_nodes, 0);
+    for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+    {
+      for (_i = 0; _i < num_nodes; ++_i)
+        _proto_flux[ph][_i] += _JxW[_qp] * _coord[_qp] * darcyQp(ph);
+    }
+  }
+  if (_pf_count_used)
+    _pf_count->setProtoFlux(_proto_flux);
```

Comment:
Kernels have the `timetepSetup` method available to them so you can at least zero. There's just no - "post step" methods that run at the end for aggregating but as long as you don't need that you should be good.

std::map is definitely slower than std::vector but so is every other container type. It's not going to kill you or anything, we use std::maps all over the place. In the case of a kernel, it's perfectly parallelized in that each processor is going to just build part of the map.

If it really does show up in timing, you can always use std::unordered_map which can be faster especially if you don't insert into it very often. It's lightning fast for lookup.